src/main/java/sparksoniq/jsoniq/runtime/iterator/postfix/PredicateIterator.java

```diff
@@ -82,9 +97,31 @@ protected void openLocal() {
         if (this._children.size() < 2) {
             throw new SparksoniqRuntimeException("Invalid Predicate! Must initialize filter before calling next");
         }
+        _filterDynamicContext = new DynamicContext(_currentDynamicContext);
+        if(_filter.getVariableDependencies().containsKey("$last"))
+        {
+            setLast();
+        }
+        if(_filter.getVariableDependencies().containsKey("$position"))
+        {
+            _position = 0;
+            _mustMaintainPosition = true;
+        }
         _iterator.open(_currentDynamicContext);
         setNextResult();
     }
+    
+    private void setLast() {
+        long last = 0;
+        _iterator.open(_currentDynamicContext);
```

Comment:
This means that the iteration on the lhs happens twice -- however, this will only actually happen if last() is called in the filter (thanks to the variable dependency check), causing no performance regression on existing queries that were not using last().